File: problems/Symmetric Tree/isSymmetric.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public bool IsSymmetric(TreeNode root) {
        if (null == root) {
            return true;
        }

        var queue = new Queue<TreeNode>();
        queue.Enqueue(root.left);
        queue.Enqueue(root.right);

        while (0 < queue.Count) {
            var node1 = queue.Dequeue();
            var node2 = queue.Dequeue();

            if (null == node1 && null == node2) continue;
            if (null == node1 || null == node2) return false;
            if (node1.val != node2.val) return false;

            queue.Enqueue(node1.left);
            queue.Enqueue(node2.right);
            queue.Enqueue(node1.right);
            queue.Enqueue(node2.left);
        }

        return true;
    }
}

File: problems/Symmetric Tree/isSymmetric.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSymmetric(TreeNode root) {
        Queue<TreeNode> bfs = new LinkedList<>();

        if (null != root) {
            bfs.add(root.left);
            bfs.add(root.right);
        }

        while (!bfs.isEmpty()) {
            TreeNode left = bfs.poll();
            TreeNode right = bfs.poll();

            if (null == left && null == right) {
                continue;
            } else if ((null == left && null != right) || (null != left && null == right) || left.val != right.val) {
                return false;
            }

            bfs.add(left.left);
            bfs.add(right.right);
            bfs.add(left.right);
            bfs.add(right.left);
        }

        return true;
    }
}
